R/distribution_formulas.R

```diff
@@ -0,0 +1,192 @@
+    if (family$link == "identity") {
+      expected <- "expected_index"
+    }
+  }
+
+  if (family[["family"]] == "multinomial") {
+    new_module <- new(TMBDmultinomDistribution)
+    expected <- "proportion_catch_numbers_at_age"
+  }
+  if (data_type == "index" | data_type == "cpue") {
+    new_module$set_observed_data(module$GetObservedIndexDataID())
+  }
+  if (data_type == "agecomp") {
+    new_module$set_observed_data(module$GetObservedAgeCompDataID())
+  }
+
+  new_module$set_distribution_links("data", module$field(expected)$get_id())
+  return(new_module)
+}
+
+#' Sets up a new distribution for a process
+#' @param par A string specifying the parameter name the distribution applies
+#'   to.
+#' @inheritParams new_data_distribution
+#' @param is_random_effect A boolean indicating whether or not the process is
+#'   estimated as a random effect.
+#' @seealso
+#' * [new_data_distribution()]
+#' @export
+new_process_distribution <- function(par,
+                                     module,
+                                     family,
+                                     sd = list(value = 1, estimated = FALSE),
+                                     is_random_effect = FALSE) {
+  families <- c("lognormal", "gaussian")
+  if (family[["family"]] == "normal") {
+    stop("use family = gaussian() instead")
+  }
+  if (!(family[["family"]] %in% families)) {
+    stop("FIMS currently does not offer this distribution for processes.")
+  }
+  if (family[["family"]] == "lognormal") {
+    new_module <- new(TMBDlnormDistribution)
+    new_module$log_logsd <- new(
+      ParameterVector,
+      log(sd$value),
+      length(sd$value)
+    )
+    new_module$log_logsd$set_all_estimable(sd$estimated)
+  }
+
+  if (family[["family"]] == "gaussian") {
+    new_module <- new(TMBDnormDistribution)
+    new_module$log_sd <- new(ParameterVector, log(sd$value), length(sd$value))
+    new_module$log_sd$set_all_estimable(sd$estimated)
+  }
+
+  module$field(par)$set_all_random(is_random_effect)
+
+  n_dim <- length(module$field(par))
+  new_module$x <- new(ParameterVector, n_dim)
+  new_module$expected_values <- new(ParameterVector, n_dim)
+  for (i in 1:n_dim) {
+    new_module$x[i]$value <- 0
+    new_module$expected_values[i]$value <- 0
+  }
+
+  new_module$set_distribution_links(
+    "random_effects",
+    module$field(par)$get_id()
+  )
+
+  return(new_module)
+}
+
+#' Lognormal family and link specification
+#'
+#' @param link A string specifying the link function for the family.
+#' @return
+#' An object of class "family" (which has a concise print method). Though this
+#' particular family has a truncated length compared to other distributions in
+#' [stats::family()].
+#' \item{family}{character: the family name.}
+#' \item{link}{character: the link name.}
+#' \item{linkfun}{function: the link.}
+#' \item{linkinv}{function: the inverse of the link function.}
+#' \item{mu.eta}{
+#'   TODO: document mu.eta
+#'   function: derivative \eqn{TODO}.
+#' }
```

Comment:
Needs more documentation.

Also, this is a REAL BIG picture comment. {sdmTMB} also has similar functions in [families.R](https://github.com/pbs-assess/sdmTMB/blob/b146eb3644321f680870a65d6e11398af0429cd8/R/families.R#L62). It would be great if we could create a package to store these where both {sdmTMB}, {FIMS}, and any other package that wants to could use them without needing to require a complex package.